src/software/ai/passing/cost_function.cpp

```diff
@@ -335,27 +296,61 @@ double getStaticPositionQuality(const Field& field, const Point& position,
 }
 
 double calculateProximityRisk(const Point& point, const Team& enemy_team,
-                              double enemy_proximity_importance)
+                              const TbotsProto::PassingConfig& passing_config)
 {
-    // Calculate a risk score based on the distance of the enemy robots from the receive
+    // Calculate a risk score based on the distance of the enemy robots from the given
     // point, based on an exponential function of the distance of each robot from the
     // receiver point
-    auto enemy_robots                 = enemy_team.getAllRobots();
-    double point_enemy_proximity_risk = 1;
-    for (const Robot& enemy : enemy_team.getAllRobots())
+    if (enemy_team.getAllRobots().empty())
     {
-        double dist = (point - enemy.position()).length();
-        point_enemy_proximity_risk *= enemy_proximity_importance * std::exp(-dist * dist);
+        return 0;
     }
-    if (enemy_robots.empty())
+
+    double risk = 0;
+    for (const Robot& enemy : enemy_team.getAllRobots())
     {
-        point_enemy_proximity_risk = 0;
+        double dist_to_enemy =
+            std::max(0.0, distance(point, enemy.position()) - ROBOT_MAX_RADIUS_METERS);
+        risk += std::exp((-dist_to_enemy * dist_to_enemy) /
+                         passing_config.enemy_proximity_importance());
     }
-    return point_enemy_proximity_risk;
+    return sigmoid(risk, 1, 2);
+}
+
+double rateKeepAwayPosition(const Point& keep_away_position, const World& world,
+                            const Pass& best_pass_so_far,
+                            const Rectangle& dribbling_bounds,
+                            const TbotsProto::PassingConfig& passing_config)
+{
+    static constexpr auto KEEPAWAY_SEARCH_CIRCLE_RADIUS = 0.5;
+
+    // the width of both the field boundary sigmoid and the circular search region sigmoid
+    static constexpr auto SIGMOID_WIDTH = 0.1;
+
+    // the region to which the optimization is (effectively) constrained to
+    Circle keepaway_search_region(world.ball().position(), KEEPAWAY_SEARCH_CIRCLE_RADIUS);
```

Comment:
This is something I wanted to updated in #3200, but I didn't have access to the right information. If we could update SensorFusion to also track the dribbling start position (or atleast a rough estimate of), then we could also update this to only search within the allowed dribbling area. I believe that keep away is the main reason (at least on master) that we end up excessive dribbling, or having to let go of the ball in dribble FSM. We could then also increase the search radius as well.

Reply:
Ok, will change

Reply:
done